Add Multiblock Simulation Chamber

eutro opened the pull request:
To do:

- [x] Decide whether to actually add this
- [x] Localization, and description
- [x] Recipe
- [x] Quest?
- [x] Pristine matter's chance tooltip in JEI: "doubled each tier" is wrong - maybe mention it

![ss](https://user-images.githubusercontent.com/11538216/113521872-344b9880-9594-11eb-9bbc-b13c0f6ec2da.png)

eutro committed:
Add Multiblock Simulation Chamber
- Missing localization, recipe and description
- Pristine matter's chance tooltip in JEI: "doubled each tier" is wrong.

eutro committed:
Remove #debug
Happens every time

Exaxxion commented:
> Decide whether to actually add this

Players having to place tons of simulation chambers in the Endgame is a common complaint, and setups involving many of them in a chunk tend to lead to both FPS (particularly when lasers are involved) and TPS lag. It's tedious to set up tons of simulation chambers and tedium isn't fun, so it seems to me like this is a good way to solve those problems.

> Localization, and description

This multiblock was based on your example structure from Multiblock Tweaker, which used the name "Stabilized Simulation Chamber." I think that's a fine name, but I'm open to suggestions.

As for description, we can come up with something reasonable for it. I'll think on this.

> Recipe

We'll need to brainstorm this. My intention is for it to be a post-tank machine (as evidenced by the blocks of Omnium in the structure). I'm considering making the recipe kind of like a tiered-up Processing Array controller with an Infinity Star taking the place of the Sunnarium.

> Quest?

We would want to add a quest for this machine to the Endgame Tab, yes. We can do that after everything else is sorted out and we're sure what the ingredients for the structure are and precisely how it will function mechanically.

> Pristine matter's chance tooltip in JEI: "doubled each tier" is wrong - maybe mention it

That's a resource string we added ourselves because we use a custom byproduct scaling function for GTCE machinery. I'm not sure if there's a way to override the tooltip strings for just this particular recipe map or not.

> [Screenshot]

We are trying to approximate the look of the simulation chamber, which has a black (or perhaps a slate grey) case with blue highlights. I definitely think it would be desirable to make the rest of the casing match the color of the bottom layer.

I understand that you were having difficulty making the controller take on the appearance of the casing material I had put in the script file I sent you the other day. You suggested in Discord that:
> if we want to make "Naquadah-Plated Dark Steel Machine Casing" or something for this then it might look better

and I think this is a good idea.

Airom42 commented:
Adding this to the discussion here to make it easier to find, this comes from a conversation in the discord. I'll also add my two cents to the matter.

> Decide whether to actually do this
I think it would be nice to have, but if the only real reason to add it is for dragon lair data, why not provide another option for dragon lair data that just requires less sim chamber spam? For example, simply providing more lair data per matter.

Doing the math, with the overclocking calculator, a  64 EU/t sim chamber recipe would overclock to 2 ticks, the original recipe being 301. This means that the chamber is equal to 150 sim chambers, which is quite worth it. However, when you input the EU/t of dragon lair data, which is 26664 EU/t, the recipe only overclocks to 39t, meaning that the multiblock is equal to a mere 7.7 sim chambers. If the multiblock produces one pristine per iteration, that goes up to the equivalent of 25.

As it stands, I'm not really sure it's worth the trouble. While setting up a bunch of dragon sims is annoying, its really cheap, whereas producing a HoTU is fairly time-consuming, and just adds to the total number you need to make for the pack. I think that if this is to be added, it should provide all of the dragon lair data needed in one structure, sort of like the RF source negates the need for any other power generation.

> name
Stabilized Sim Chamber works, maybe something like Simulation Chamber Accelerator could also make sense, depending on the theory of how the thing works - is it a super-powered sim chamber, or does it overcharge the sim chamber placed inside of it?

Exaxxion commented:
@Airom42 I think you'll find you're off on the math. The conversion is 4 RF : 1 EU, so I think you got it backwards.
```zs
    "dragon":           6666,
```
Means that the dragon lair data operates at a base energy of 1666 EU/t (integer division drops the decimal). Lair data would thus operate [on these parameters](https://omnifactorydevs.github.io/OverclockingCalculator/#eut=1666&duration=301). At MAX tier, it should overclock down to 2t per operation, for an average of 3 pristines per second at Self Aware. I think this is absolutely worth the investment.

eutro commented:
> Localisation

I think I will go for "Stabilized Simulation Chamber" then, and we'll worry about a description later.

> Recipe

I think a HotU would be excessive: this doesn't offer too much more than a bunch of regular simulation chambers, with the obvious benefit being less tedious setup. The omnium requirement would place it squarely post-tank, where mass-producing simulation chambers and data models isn't really an obstacle: you'd be paying the HotU just to avoid the grind of setting up a bunch of simulation chambers. A better recipe could involve an assortment of self-aware data models, perhaps.

> JEI Tooltip

There's no existing way to change which localisation key a set of recipes uses. Perhaps something like `RecipeMap.chanceTooltipFunction` could be PR-ed to GTCE.

> Multiblock Structure

I'll go ahead and add the "Naquadah-Plated Dark Steel Machine Casing" then.

smudgerox commented:
> @Airom42 I think you'll find you're off on the math. The conversion is 4 RF : 1 EU, so I think you got it backwards.
> 
> ```zenscript
>     "dragon":           6666,
> ```
> 

This math error was totally my fault. Was too tired when him and I were discussing this. 

Anyways, after thinking about it more, I agree with Eutro. HOTU does feel a bit excessive after playing with it earlier. It should be cheap in the way the PA is where it's just trivial to spam when you get to the point where these are needed. 

One thing I was thinking about earlier is that with this, you'll need to spam loot fabs now. I propose letting this multiblock perform the function of the loot fab as well, and use a integrated circuit to determine the output (much like Neeve's Stabilized Miners in STE). 

As for the name, I think something like "Simulation Supercomputer" might be a cool name. I don't really think the Stabilized thing makes sense, since it's not really "unstable" to begin with. 

Lastly, I think the blocks used for the decoration on the sides could use some change. 

The left side (when facing the front) is backwards, and the coal block is honestly kinda ugly. I came up with a couple alternative blocks that could be used, if you're open to changing it.

eutro commented:
"Simulation Supercomputer" does sound a lot better. And yeah, the coal blocks could certainly be improved.

Airom42 commented:
> This math error was totally my fault. Was too tired when him and I were discussing this.
Overclocking to those proper numbers seems much more reasonable. Good catch!

> I think a HotU would be excessive:
I think it depends on how many of these you need to build. If you only need to build one and it solves all your lair data needs, then a HoTU is worth it, but if you need to make multiple, then it starts to get expensive.

> name
I also like Simulation Supercomputer

UhahaUhaha commented:
Name: "Deep Simulation Factory" by tanjiro#0577

Exaxxion commented:
I like the sound of Simulation Supercomputer as well.

Processing Arrays have a technology gate of Sunnarium, requiring the player to both progress through most or all of Fission and create an IV alloy smelter to make them at all. Sunnarium is a slow material to produce when you first unlock it, and a limiting factor for how many PAs you can make.

I was thinking of what would make a good endgame equivalent to this and the Infinity Catalyst seemed to fit the best. All of the other materials are trivial the moment you reach tank, so there would not be much sense of accomplishment in unlocking this structure. If you can think of something besides HotU-based materials that fits this bill, I'm willing to consider it.

We've also made several changes in dev over the past months which reduce the pain of Dragon Lair Data, including switching the crafting recipe from requiring hellish matter (which, frankly, players don't tend to have all that much of) to extraterrestrial matter (which players are usually drowning in). You probably aren't going to need to make more than a few of these things ever, because of how potent they are.

As for the suggestion of using a circuit to have them output specific items directly, I'll have to think about it some more. Creating all the recipes becomes a lot more complicated if we do that and we'd want to include the Loot Fabricator's operation time in the recipe duration.

It takes the expected value of ~19 simulation chambers operating on one Self Aware model to saturate the processing speed of one loot fabricator. Going by a previously mentioned figure of this thing being essentially 250 simulation chambers, you'd need 13 loot fabricators to keep up with that.

Exaxxion commented:
When you have some time @eutro I would appreciate if you could continue this, as it's an important enhancement we'd like to get in 1.3.

We had talked about creating a new casing material "Naquadah-Plated Dark Steel Machine Casing" to use that would look more the part and that way we would avoid things like coal blocks. I believe we also opted to rename this to "Simulation Supercomputer."

I'm not too worried about the specifics of the controller recipe, we can tweak that for balance as needed. I also don't think it is necessary to merge the functionality of the loot fabricator into this structure. The saturation ratio is already considerable at 19:1 and building one of these and roughly ten loot fabricators is far less ridiculous than building hundreds of Simulation Chambers.

ALongStringOfNumbers committed:
Add contenttweaker casing block

ALongStringOfNumbers committed:
Add temporary recipe for the controller

ALongStringOfNumbers commented:
I have added in a contenttweaker casing for the Naquadah-plated dark steel casing discussed above. It is currently using the microversium texture until an actual texture is obtained for it.
It also has a temporary recipe, which can be debated over.

I have also added localization for the controller, as discussed in the issue above, and for the multiblock description. The multiblock description could use some more refinement, but I just made one to get something in there.

In addition, I have added a temporary recipe for the controller, which places it in the late game, but does not gate it behind a heart of the universe.

I have not added a quest for this, as it would conflict with the ongoing questbook PR, but it does really need its own quest explaining the mechanics of how it works, and addressing the tooltip issue, as that cannot be adjusted on a per recipemap basis.

ALongStringOfNumbers committed:
localize recipemap name

ALongStringOfNumbers commented:
Also a suggestion I would make is have the recipes take more EU/t depending on the tier of the model. It doesn't necessarily have to go to the next tier every model, but some increase in EU/T for the higher pristine chance would work well in my opinion.

ALongStringOfNumbers committed:
Adjust structure layout, add new casing

johnch18 commented:
Uses CE Nt plates, not the More Plates ones. Needs to be replaced or oredicted.

ALongStringOfNumbers committed:
Switch to oredict Neutronium plate

johnch18 commented:
https://pastebin.com/ChrZSzbt
A tool for analyzing the multi at different overclocks.

johnch18 commented:
Thermal Elemental models don't work in the multi

ALongStringOfNumbers committed:
Add Thermal Elemental model

Cyber1551 commented:
From what I understand from messing around with it you can only use 1 data model per supercomputer? Maybe I'm doing it wrong but a suggestion would be to run all data models in the input bus at once. Seems more reasonable for the size/complexity.

eiis1000 commented:
This PR is on the 1.3 roadmap but there's been no activity in a month. At the current point what's missing for it to be full featured? If I have some time I might try to fill in some gaps.

Exaxxion commented:
It's mainly just that we've been focusing on other issues this past month, namely #665 and #757 (we also maintain SoG and just finished a major PR there last week). 

As I understand it, what's left is to fix some merge conflicts in the files, finalize the controller recipe, and create a quest for it. None of those things are particularly easy for someone else to assist with.

ALongStringOfNumbers committed:
Merge branch 'dev' into multiblock-simulation-chamber

Exaxxion committed:
Change controller recipe

Require HotU to make it obvious this multiblock is basically a creative
tier item, intended to be made (and not trivially spammable) post-tank.

Exaxxion committed:
Merge remote-tracking branch 'origin/dev' into multiblock-simulation-chamber

Exaxxion committed:
Add quest for the Simulation Supercomputer

Exaxxion commented:
Added a tentative recipe for the controller cementing that this is a creative-tier structure. Caught the branch up to date, and added a quest for the Simulation Supercomputer branching off of the Creative Tank quest. Ideally players will consider making this with their second HotU (should be able to make enough lair data for that before feeling compelled to spam hundreds of sim chambers like they currently do).

The quest text clarifies that the pristine chance is the same as you would have in the data model run in a Simulation Chamber, it just can overclock the speed of the computations. The universal chance tooltip is a limitation we can't really do much about without changes in GTCE.

NotMyWing commented on overrides/resources/contenttweaker/lang/en_us.lang:
```diff
@@ -232,3 +232,6 @@ item.contenttweaker.hand_framing_tool.desc6=The tool can then be used to convert
 item.contenttweaker.pulsatingmesh.name=Pulsating Mesh
+
+tile.contenttweaker.simulation_casing.name=Naquadah-Plated Dark Steel Machine Casing
+tile.contenttweaker.simulation_casing_2.name=Naquadah Casing
```
Nitpick: should be "Naquadah Machine Casing" to be in line with other machine casings.

Exaxxion replied:
Good point, I'll make that change.

NotMyWing commented on overrides/scripts/MultiblockSimulationChamber.zs:
```diff
@@ -0,0 +1,280 @@
+assembler.recipeBuilder()
+    .inputs(<metaitem:plateNaquadah> * 4)
+    .outputs(<contenttweaker:simulation_casing_2>)
+    .duration(200).EUt(30).buildAndRegister();
```
Nitpick: possible conflict with naquadah rotors.
Not that anyone would ever use them though.

Exaxxion replied:
I'm guessing you mean Turbine Blades? Those uniquely use Circuit 10 in the recipe though so there should be extremely little risk of recipe overlap.

NotMyWing replied:
Issues may come from the fact that assembling machines may erroneously pick the naquadah casing recipe when they're supposed to be making blades, akin to #350. But then again, I don't think anyone would ever make naquadah blades, so it's mostly just a nitpick.

Exaxxion replied:
I think it's not really any more of a problem than the 6 plate + 1 frame machine casing recipes from GTCE that don't use a disambiguating circuit. Those could cause problems in theory if you tried to use a Circuit 8 in it or something, but blocking mode largely prevents such issues (and it's completely avoidable by not trying to run possibly conflicting recipes from the same machine).

Exaxxion committed:
Rename Naquadah casing to Naquadah Machine Casing
